File: python_template/src/load.py
import os
import sys
sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__))))

import json
import logging
from Inspector import *
import time
import boto3
import csv
import pymysql


def load_write(f, dblink, dbname, tablename, username, password, inspector):
    try:
        con = pymysql.connect(host=dblink, db=dbname, user=username, passwd=password)

        cursor = con.cursor()
        # drop tabel if exists
        cursor.execute("DROP TABLE IF EXISTS " + tablename) 
        # create table
        cursor.execute("CREATE TABLE " + tablename + " (Region VARCHAR(50), Country VARCHAR(50), `Item Type` VARCHAR(50), `Sales Channel` VARCHAR(50), `Order Priority` VARCHAR(50),`Order Date` VARCHAR(50), `Order ID` INT PRIMARY KEY AUTO_INCREMENT, `Ship Date` VARCHAR(50), `Units Sold` DOUBLE, `Unit Price` DOUBLE, `Unit Cost` DOUBLE, `Total Revenue` DOUBLE, `Total Cost` DOUBLE, `Total Profit` DOUBLE, `Gross Margin` DOUBLE, `Order Processing Time` DOUBLE ) ")
        # insert data
        sql = ("INSERT INTO "+ tablename + " (Region, Country, `Item Type` , `Sales Channel` , `Order Priority` ,`Order Date`,  `Order ID`, `Ship Date`, `Units Sold`, `Unit Price`, `Unit Cost`, `Total Revenue`, `Total Cost`, `Total Profit`, `Gross Margin`, `Order Processing Time`) VALUES (%s, %s, %s, %s, %s,%s, %s, %s, %s, %s,%s, %s, %s, %s, %s, %s)")
        next(f)
        for line in f:
            line = [None if cell == '' else cell for cell in line]
            cursor.execute(sql, line)
            
        con.commit()
        cursor.close()
    except pymysql.MySQLError as e:
        inspector.addAttribute("error", "ERROR: Unexpected error: Could not connect to MySQL instance.")
        inspector.addAttribute("stackTrace", e)

def handler(request, context):
    # Import the module and collect data
    inspector = Inspector()

    key = str(request['key'])
    bucketname = str(request['bucketname'])
    dblink = str(request['dblink'])
    dbname = str(request['dbname'])
    tablename = str(request['tablename'])
    username = str(request['username'])
    password = str(request['password'])

    #read csv from s3
    s3 = boto3.client('s3')
    csv_file = s3.get_object(Bucket=bucketname, Key=key)
    f = csv.reader(csv_file["Body"].read().decode('utf-8').split('\n')[:-1])
    load_write(f, dblink, dbname, tablename, username, password, inspector)

    return inspector.finish()


if __name__ == "__main__":
    context = None
    request = {}
    request["key"] = "Processed_100_Sales_Records.csv"
    request["bucketname"] = "test.bucket.562f20.leb"
    request["dblink"] = "local"
    request["dbname"] = "tlq562"
    request["tablename"] = "562_test"
    request["username"] = "root"
    request["password"] = "password"

    response = handler(request, context)
    print(response)

File: java_template/src/main/java/lambda/Load.java
package lambda;

import com.amazonaws.services.lambda.runtime.Context;
import com.amazonaws.services.lambda.runtime.RequestHandler;
import com.amazonaws.services.s3.AmazonS3ClientBuilder;
import com.amazonaws.services.s3.AmazonS3;
import com.amazonaws.services.s3.model.S3Object;
import com.amazonaws.services.s3.model.GetObjectRequest;
import saaf.Inspector;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.sql.Connection;
import java.util.Properties;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.SQLException;
import org.apache.commons.csv.CSVParser;
import org.apache.commons.csv.CSVRecord;
import org.apache.commons.csv.CSVFormat;


public class Load implements RequestHandler<HashMap<String, Object>, HashMap<String, Object>> {

    public CSVParser readCSV(
        Inspector inspector, 
        AmazonS3 s3Client, 
        String bucketName, 
        String fileName
    ) {
        S3Object s3Object = s3Client.getObject(new GetObjectRequest(bucketName, fileName));
        InputStream csvInputStream = s3Object.getObjectContent();
        InputStreamReader csvInputStreamReader = new InputStreamReader(csvInputStream);
        CSVParser csvParser = null;
        try {
            csvParser = CSVFormat.EXCEL.withHeader().parse(csvInputStreamReader);
        } catch (IOException e) {
            inspector.addAttribute("error", "Error while reading CSV file from S3");
            inspector.addAttribute("stackTrace", e.getStackTrace());
        }

        return csvParser;
    }

    public void writeSQL(
        Inspector inspector, 
        CSVParser csvParser,
        String dbLink, 
        String dbName, 
        String username, 
        String password, 
        String tableName
    ) {
        Properties properties = new Properties();
        properties.setProperty("user", username);
        properties.setProperty("password", password);
        properties.setProperty("useSSL", "false");

        int batchSize = 1;

        try {
            Connection connection = DriverManager.getConnection("jdbc:mysql://" + dbLink + ":3306/" + dbName, properties);
            PreparedStatement preparedStatement = connection.prepareStatement("DROP TABLE IF EXISTS `" + tableName + "`;");
            preparedStatement.execute();
            preparedStatement = connection.prepareStatement(
                "CREATE TABLE " + tableName
                + " (Region VARCHAR(40), Country VARCHAR(40), `Item Type` VARCHAR(40), `Sales Channel` VARCHAR(40),`Order Priority` VARCHAR(40), `Order Date` VARCHAR(40),`Order ID` INT PRIMARY KEY, `Ship Date` VARCHAR(40), `Units Sold` INT,`Unit Price` DOUBLE, `Unit Cost` DOUBLE, `Total Revenue` DOUBLE, `Total Cost` DOUBLE, `Total Profit` DOUBLE, `Order Processing Time` INT, `Gross Margin` FLOAT) ENGINE = MyISAM;"
            );
            preparedStatement.execute();
            String insertSQL = "INSERT INTO " + tableName 
                + " (Region, Country, `Item Type`, `Sales Channel`, `Order Priority`, `Order Date`, `Order ID`, `Ship Date`, `Units Sold`, `Unit Price`, `Unit Cost`, `Total Revenue`, `Total Cost`, `Total Profit`, `Order Processing Time`, `Gross Margin`) values (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ? ,? ,? ,?)";
            PreparedStatement statement = connection.prepareStatement(insertSQL);

            // Create header
            ArrayList<String> header = new ArrayList<String>();
            header.add("Region");
            header.add("Country");
            header.add("Item Type");
            header.add("Sales Channel");
            header.add("Order Priority");
            header.add("Order Date");
            header.add("Order ID");
            header.add("Ship Date");
            header.add("Units Sold");
            header.add("Unit Price");
            header.add("Unit Cost");
            header.add("Total Revenue");
            header.add("Total Cost");
            header.add("Total Profit");
            header.add("Order Processing Time");
            header.add("Gross Margin");
            
            List<CSVRecord> csvList;
            csvList = csvParser.getRecords();
            for (int row_i=0; row_i<csvList.size(); row_i++) {
                CSVRecord csvRecord = csvList.get(row_i);
                for (int key_i=0; key_i<header.size(); key_i++) {
                    String key = header.get(key_i);
                    statement.setString(key_i + 1, csvRecord.get(key));
                }
                statement.addBatch();
                if ((row_i+1) % batchSize == 0) {
                    statement.executeBatch();
                }
            }
            statement.executeBatch();
            statement.close();
            connection.close();
        } catch (SQLException | IOException e) {
            inspector.addAttribute("error", "Error while writing SQL to database");
            inspector.addAttribute("stackTrace", e.getStackTrace());
        }
    }

    public HashMap<String, Object> handleRequest(HashMap<String, Object> request, Context context) {
        
        Inspector inspector = new Inspector();

        AmazonS3 s3Client = AmazonS3ClientBuilder.standard().build();
        String bucketName = (String) request.get("bucketName");
        String fileName = (String) request.get("fileName");
        String dbLink = (String) request.get("dbLink");
        String dbName = (String) request.get("dbName");
        String username = (String) request.get("username");
        String password = (String) request.get("password");
        String tableName = (String) request.get("tableName");

        CSVParser csvParser = readCSV(inspector, s3Client, bucketName, fileName);
        writeSQL(inspector, csvParser, dbLink, dbName, username, password, tableName);

        return inspector.finish();
    }
}
